includes/admin/extensions/abstract-extension.php

```diff
@@ -0,0 +1,269 @@
+	 * @param int    $item_id      Optional: the item ID.
+	 * @return array
+	 */
+	protected function get_button_parameters( $product_data, $item_id = false ) {
+		if ( empty( $item_id ) ) {
+			$item_id = $this->item_id;
+		}
+		$body   = $this->get_api_body();
+		$type   = $this->get_type( $body );
+		$id     = ! empty( $body[ $type ] ) ? $body[ $type ] : $this->item_id;
+		$button = array(
+			'type'    => $type,
+			'id'      => $id,
+			'product' => $item_id,
+		);
+		// If the extension is not installed, the button will prompt to install and activate it.
+		if ( ! $this->manager->is_plugin_installed( $product_data['basename'] ) ) {
+			if ( $this->manager->pass_can_download() ) {
+				$button = array(
+					/* translators: The extension name. */
+					'button_text' => sprintf( __( 'Log In to Your Account to Download %s', 'easy-digital-downloads' ), $product_data['title'] ),
+					'href'        => $this->get_upgrade_url( $product_data, $item_id, true ),
+					'new_tab'     => true,
+					'type'        => $type,
+				);
+			} else {
+				$button = array(
+					/* translators: The extension name. */
+					'button_text' => sprintf( __( 'Upgrade Today to Access %s!', 'easy-digital-downloads' ), $product_data['title'] ),
+					'href'        => $this->get_upgrade_url( $product_data, $item_id ),
+					'new_tab'     => true,
+					'type'        => $type,
+				);
+			}
+		} elseif ( ! $this->manager->is_plugin_active( $product_data['basename'] ) ) {
+			// If the extension is installed, but not activated, the button will prompt to activate it.
+			$button['plugin'] = $product_data['basename'];
+			$button['action'] = 'activate';
+			/* translators: The extension name. */
+			$button['button_text'] = sprintf( __( 'Activate %s', 'easy-digital-downloads' ), $product_data['title'] );
+		}
+
+		return $button;
+	}
+
+	/**
+	 * Gets the upgrade URL for the button.
+	 *
+	 * @todo add UTM parameters
+	 * @since 2.11.x
+	 * @param array $product_data The array of provided data about the extension.
+	 * @param int   $item_id      The item/product ID.
+	 * @param bool  $has_access   Whether the user already has access to the extension (based on pass level).
+	 * @return string
+	 */
+	private function get_upgrade_url( $product_data, $item_id, $has_access = false ) {
+		$url            = 'https://easydigitaldownloads.com';
+		$tab            = ! empty( $_GET['tab'] ) ? sanitize_text_field( $_GET['tab'] ) : '';
+		$utm_parameters = array(
+			'p'            => urlencode( $item_id ),
+			'utm_source'   => 'settings',
+			'utm_medium'   => urlencode( $tab ),
+			'utm_campaign' => 'admin',
+			'utm_term'     => urlencode( $product_data['slug'] ),
+		);
+
+		if ( $has_access ) {
+			$url = 'https://easydigitaldownloads.com/your-account/your-downloads/';
+			unset( $utm_parameters['p'] );
+		} elseif ( ! empty( $product_data['upgrade_url'] ) ) {
+			$url = esc_url( $product_data['upgrade_url'] );
```

Comment:
```suggestion
			$url = esc_url_raw( $product_data['upgrade_url'] );
```

I think this is more appropriate. `esc_url()` would be good on display (which I think you're already doing). `esc_url_raw()` is more of a sanitize function.